### DISCUSSION TITLE
v1.2.0 stable

### DISCUSSION THREAD

**Discussion by tomhea** [tomhea/flip-jump#236] [Announcements]
Version 1.2.0 changes:
- Flipjump is now a python package. `pip install flipjump`:
  - Run the new command line tool: `fj hello-world.fj`.
  - Import from python: `from flipjump import assemble_and_run, ...`.
  - Poetry support via the new pyproject.toml.
- Project reorganization: the flipjump python source files are under `flipjump/` (reorganized and split into smaller files), and the stl files are under `flipjump/stl/`.
- The flipjump_quickstart.py file: easy access to the main flipjump features from Python.
- Improved READMEs and documentation.

It is all tested, and ready to be used on windows or linux (and mac shouldn't be a problem), with python 3.7+.
_The only exception is hex.div which still doesn't work (its tests are xfailed), so you can use the slower-but-stable bit.div+casts for division._

<hr /><em>This discussion was created from the release <a href='https://github.com/tomhea/flip-jump/releases/tag/1.2.0'>v1.2.0 stable</a>.</em>

**Comment by tomhea** [comment]
# Update - flipjump is now a python package !
It's much easier to use now - use the new cli tool: `fj -h`, or just `import flipjump` :)

@guym0r @coolslutguy420 @ivelieu @lestrozi @vxgmichel @culverit @hmltn2 @jaripekkary @mtve @ebigram @emenel @nst @roma-glushko @Siss3l @tuxcanfly @elhadjx @Peefy @fabiosvm @andrew-johnson-4 @mgmalheiros @frediff @HaraldKorneliussen @adamharmansky @dfischer @CCXXXI @egon-r @Sophia-Cristina @creaktive @Lyxal @parched @TonikHorkel @ivannz @FrogBoy2008 @R-Jones @neauoire @idoherman @danytnt @conilas @gavinbm @jakiki6 @Taardisaa @Clinale @zephryn @thisisclosure @kamil-adam @Polarzz @Sherlockcxk @idozo @betedgi

**Comment by vxgmichel** [comment]
That's great! :tada:

I can now use `flipjump.fjm.fjm_reader.Reader` to [parse `.fjm` files](https://github.com/vxgmichel/marbles/blob/11c5eee97312980d8a551170fede13f03a38dfa2/flipjump-to-marbles.py#L123-L126):
```python
        reader = FjmReader(program)
        width = reader.memory_width
        size = cast(int, max(reader.memory)) + 1
        words = [reader.memory[i] for i in range(size)]
```
@tomhea Can you confirm that this interface will remain stable?

**Reply by tomhea** [reply]
@vxgmichel Hi!
The .memory isnt stable. Use the stable .get_word().

Also the way you retrieve the size isn't stable (i will make the memory private soon), but i don't think there is a good stable way of getting the "size". Ill add one soon (~two weeks time), and ill patch things out in a future fj version 1.2.1.
For now you can retrieve the size as you do, and ill update here if the api will change in 1.2.1.

Tell me if you need anything else :)

**Reply by vxgmichel** [reply]
> Use the stable .get_word().

It seems like this API is more designed to be used when running the program since it takes a bit address as an argument, and it includes this "garbage handling" logic that is not useful to me.

What I'm looking for is simply a way to get the full memory as a list of words from an `fjm` file. Would you mind including a stable `reader.get_memory()` helper or something similar? That'd be great :)

**Reply by tomhea** [reply]
Its a bit more complicated than that, as the memory is not always just a list of values, as the memory is not always continuous.

With the `segment` / `reserve` ops you'll get different segments in memory, not likely to be adjacent to one another, and some of them might be reserved with 0's. Some of them might be filled with a different fixed value (maybe in a new fjm version).

How would you suggest to handle that? As I'd like to add that `get_memory()` for you, and I even think it'll be used not just by you.

I thought of a default_dict, from address to value, (for stuff like reserved 0s), but will it be enough for you? What would you suggest?

**Reply by vxgmichel** [reply]
A `defaultdict` sounds great! In my specific case I need the following information:

- the memory width
- the memory size (or at least the maximum address that the program is meant to access)
- a mapping to get a word value from a word address

As long as this info is exposed through a public API I'll be happy :)

**Reply by tomhea** [reply]
Cool, ill think how and implement it, and ill do my best to update here :)

**Reply by tomhea** [reply]
Hi, I haven't forgot you, though I published v1.2.1 now, I'll solve the #238 issue I opened for that matter probably in v1.2.2.

The API haven't changed in this version.

**Reply by tomhea** [reply]
@vxgmichel 
Done (Now in main branch, and will enter in version 1.3.0). The stable interface is now:
```python
reader = FjmReader(program)
width = reader.memory_width
segments = reader.memory_segments  
# If you have only one segment, you can do "size  = segments.segment_start + segments.segment_length - 1"
memory = reader.get_memory()  # defaultdict
words = [memory[i] for i in range(size)]
```
